File: packages/cli/src/protocol.ts
import type { ProtocolToolApproval } from "./protocol/approval.js";
import {
  serializeEvent,
  serviceCapabilities,
} from "./protocol/capabilities.js";
import { RunRegistry } from "./protocol/run-registry.js";
import { jsonRpcMessage } from "./protocol/wire.js";
import {
  RUNTIME_PACKAGE_NAME,
  RUNTIME_VERSION,
} from "@truss-harness/runtime";
import {
  LOCAL_SERVICE_PROTOCOL_VERSIONS,
  type RuntimeServiceCapabilities,
  type RuntimeServiceErrorCode,
  type RuntimeServiceEventSource,
  type RuntimeServiceHost,
  type RuntimeServiceMessage,
  type RuntimeServiceResult,
  type RuntimeServiceRuntime,
  type RuntimeServiceWireMessage,
} from "./protocol-contracts.js";
import {
  protocolObject,
  requestError,
  sanitizeAttachments,
  sanitizeContext,
  sanitizeMessages,
  sanitizePrompt,
  validRequestId,
} from "./protocol-validation.js";

export { ProtocolToolApproval } from "./protocol/approval.js";
export * from "./protocol-contracts.js";
export { validateRuntimeServiceHandshake } from "./protocol/compatibility.js";
export {
  activateRuntimeHost,
  installRuntimeHostArtifact,
  parseRuntimeHostManifest,
  readRuntimeHostActivation,
  rollbackRuntimeHost,
  verifyRuntimeHostArtifact,
  runtimeHostStorePath,
} from "./runtime-delivery.js";
export type {
  RuntimeHostActivation,
  RuntimeHostManifest,
} from "./runtime-delivery.js";

export interface RuntimeServiceOptions {
  readonly runtime: RuntimeServiceRuntime;
  readonly events: RuntimeServiceEventSource;
  readonly write: (message: RuntimeServiceWireMessage) => void;
  readonly approval?: ProtocolToolApproval;
  readonly host?: RuntimeServiceHost;
  readonly serverVersion?: string;
  readonly capabilities?: Partial<RuntimeServiceCapabilities>;
  /** Kept during migration so released VS Code clients can still connect. */
  readonly allowLegacyRequests?: boolean;
}

/**
 * Stateful, transport-independent local service. Clients exchange one JSON
 * object per line; the class owns request routing and deterministic cleanup.
 */
export class RuntimeService {
  private readonly runs: RunRegistry;
  private readonly unsubscribe: () => void;
  private readonly unsubscribeApproval?: () => void;
  private readonly capabilities: RuntimeServiceCapabilities;
  private initialized = false;
  private closed = false;
  private wireMode: "legacy" | "jsonrpc" | undefined;

  constructor(private readonly options: RuntimeServiceOptions) {
    this.capabilities = serviceCapabilities(options);
    this.runs = new RunRegistry({
      runtime: options.runtime,
      send: (message) => this.send(message),
      sendError: (code, message, requestId) =>
        this.sendError(code, message, requestId),
      denySession: (sessionId) => options.approval?.denySession(sessionId),
      isClosed: () => this.closed,
    });
    this.unsubscribe = options.events.subscribe((event) => {
      const requestId = this.runs.requestForSession(event.sessionId);
      if (!requestId || this.closed) return;
      this.send({
        type: "event",
        requestId,
        event: serializeEvent(event),
      });
    });
    this.unsubscribeApproval = options.approval?.subscribe((call, session) => {
      const requestId = this.runs.requestForSession(session.id);
      if (!requestId || this.closed) return;
      this.send({
        type: "approval_request",
        requestId,
        sessionId: session.id,
        callId: call.id,
        tool: call.name,
        input: call.input,
      });
    });
  }

  async handleLine(line: string): Promise<"continue" | "shutdown"> {
    if (this.closed) return "shutdown";
    let value: unknown;
    try {
      value = JSON.parse(line) as unknown;
    } catch {
      this.sendError(
        "invalid_json",
        "Invalid JSON request. Send one JSON object per line.",
      );
      return "continue";
    }

    let request = protocolObject(value);
    if (request?.jsonrpc === "2.0") {
      if (this.wireMode === "legacy") {
        this.sendError(
          "invalid_request",
          "A connection cannot mix legacy and JSON-RPC messages.",
          typeof request.id === "string" ? request.id : undefined,
        );
        return "continue";
      }
      this.wireMode = "jsonrpc";
      request = this.jsonRpcRequest(request);
    } else {
      if (this.wireMode === "jsonrpc") {
        this.sendError(
          "invalid_request",
          "A connection cannot mix JSON-RPC and legacy messages.",
          typeof request?.requestId === "string"
            ? request.requestId
            : undefined,
        );
        return "continue";
      }
      this.wireMode = "legacy";
    }
    const requestId = request?.requestId;
    if (!request || typeof request.type !== "string") {
      this.sendError("invalid_request", "A service request needs a type.");
      return "continue";
    }
    if (!validRequestId(requestId)) {
      this.sendError(
        "invalid_request",
        "A service request needs a non-empty requestId.",
      );
      return "continue";
    }

    if (request.type === "initialize") {
      this.initialize(requestId, request);
      return "continue";
    }
    if (
      !this.initialized &&
      (this.wireMode === "jsonrpc" ||
        this.options.allowLegacyRequests === false)
    ) {
      this.sendError(
        "invalid_request",
        "Initialize the local service before sending requests.",
        requestId,
      );
      return "continue";
    }

    try {
      if (request.type === "ping") {
        this.send({
          type: "response",
          requestId,
          result: { pong: true },
        });
        return "continue";
      }
      if (request.type === "shutdown") {
        this.send({
          type: "response",
          requestId,
          result: { shutdown: true },
        });
        return "shutdown";
      }
      if (request.type === "create_session") {
        await this.createSession(requestId, request.messages);
        return "continue";
      }
      if (request.type === "run") {
        this.startRun(requestId, request);
        return "continue";
      }
      if (request.type === "abort") {
        this.cancel(requestId, true);
        return "continue";
      }
      if (request.type === "cancel") {
        if (!validRequestId(request.targetRequestId))
          throw new Error("A cancel request needs a targetRequestId.");
        const cancelled = this.cancel(request.targetRequestId, false);
        this.send({
          type: "response",
          requestId,
          result: {
            cancelled,
            targetRequestId: request.targetRequestId,
          },
        });
        return "continue";
      }
      if (request.type === "tool_approval") {
        if (
          typeof request.callId !== "string" ||
          typeof request.approved !== "boolean"
        )
          throw new Error(
            "A tool approval needs a callId and approved boolean.",
          );
        if (!this.options.approval?.resolve(request.callId, request.approved))
          this.sendError(
            "unknown_request",
            "No tool approval is pending for that call.",
            requestId,
          );
        else if (this.wireMode === "jsonrpc")
          this.send({
            type: "response",
            requestId,
            result: { approvalResolved: true },
          });
        return "continue";
      }
      if (request.type === "test_provider") {
        const testProviderConnection =
          this.options.host?.testProviderConnection;
        if (!testProviderConnection)
          return this.unavailableCapability(
            requestId,
            "Provider connection testing",
          );
        await this.hostResponse(requestId, async () => ({
          providerConnection: await testProviderConnection(),
        }));
        return "continue";
      }
      if (request.type === "list_profiles") {
        const listProfiles = this.options.host?.listProfiles;
        if (!listProfiles)
          return this.unavailableCapability(requestId, "Profile discovery");
        await this.hostResponse(requestId, async () => ({
          profiles: await listProfiles(),
        }));
        return "continue";
      }
      if (request.type === "mcp_status") {
        const listMcpServers = this.options.host?.listMcpServers;
        if (!listMcpServers)
          return this.unavailableCapability(requestId, "MCP status");
        await this.hostResponse(requestId, async () => ({
          mcpServers: await listMcpServers(),
        }));
        return "continue";
      }
    } catch (error) {
      this.sendError("invalid_request", requestError(error), requestId);
      return "continue";
    }

    this.sendError(
      this.wireMode === "jsonrpc" ? "method_not_found" : "invalid_request",
      `Unknown service request type: ${request.type}`,
      requestId,
    );
    return "continue";
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    this.unsubscribe();
    this.unsubscribeApproval?.();
    await this.runs.close();
    this.options.approval?.denyAll();
  }

  private initialize(
    requestId: string,
    request: Record<string, unknown>,
  ): void {
    if (
      !Array.isArray(request.protocolVersions) ||
      !request.protocolVersions.every(
        (version) => Number.isInteger(version) && version > 0,
      )
    ) {
      this.sendError(
        "invalid_request",
        "Initialize requires a protocolVersions array.",
        requestId,
      );
      return;
    }
    const protocolVersion = [...request.protocolVersions]
      .sort((left, right) => Number(right) - Number(left))
      .find((version) =>
        LOCAL_SERVICE_PROTOCOL_VERSIONS.includes(
          version as (typeof LOCAL_SERVICE_PROTOCOL_VERSIONS)[number],
        ),
      );
    if (!protocolVersion) {
      this.send({
        type: "error",
        requestId,
        code: "unsupported_protocol",
        message:
          "The client and local service have no shared protocol version.",
        supportedProtocolVersions: LOCAL_SERVICE_PROTOCOL_VERSIONS,
      });
      return;
    }
    this.initialized = true;
    this.send({
      type: "response",
      requestId,
      result: {
        protocolVersion,
        server: {
          name: "truss-cli",
          version: this.options.serverVersion ?? "development",
          identity: {
            runtime: {
              packageName: RUNTIME_PACKAGE_NAME,
              version: RUNTIME_VERSION,
            },
            protocolVersions: LOCAL_SERVICE_PROTOCOL_VERSIONS,
          },
        },
        capabilities: this.capabilities,
      },
    });
  }

  private jsonRpcRequest(
    request: Record<string, unknown>,
  ): Record<string, unknown> | undefined {
    if (!validRequestId(request.id) || typeof request.method !== "string") {
      this.sendError(
        "invalid_request",
        "A JSON-RPC request needs a string id and method.",
        typeof request.id === "string" ? request.id : undefined,
      );
      return undefined;
    }
    const params =
      request.params === undefined ? {} : protocolObject(request.params);
    if (!params) {
      this.sendError(
        "invalid_request",
        "JSON-RPC params must be an object.",
        request.id,
      );
      return undefined;
    }
    const common = { ...params, requestId: request.id };
    if (request.method === "initialize")
      return { ...common, type: "initialize" };
    if (request.method === "session/create")
      return { ...common, type: "create_session" };
    if (request.method === "run/start") return { ...common, type: "run" };
    if (request.method === "run/cancel") return { ...common, type: "cancel" };
    if (request.method === "approval/resolve")
      return { ...common, type: "tool_approval" };
    if (request.method === "provider/test")
      return { ...common, type: "test_provider" };
    if (request.method === "profiles/list")
      return { ...common, type: "list_profiles" };
    if (request.method === "mcp/status")
      return { ...common, type: "mcp_status" };
    if (request.method === "service/ping") return { ...common, type: "ping" };
    if (request.method === "service/shutdown")
      return { ...common, type: "shutdown" };
    return { ...common, type: request.method };
  }

  private async createSession(
    requestId: string,
    messages: unknown,
  ): Promise<void> {
    const sanitized = sanitizeMessages(messages);
    try {
      const session = await this.options.runtime.createSession(sanitized);
      this.send({
        type: "response",
        requestId,
        result: { sessionId: session.id },
      });
    } catch (error) {
      this.sendError("internal_error", requestError(error), requestId);
    }
  }

  private startRun(requestId: string, request: Record<string, unknown>): void {
    const prompt = sanitizePrompt(request.prompt);
    if (
      request.sessionId !== undefined &&
      typeof request.sessionId !== "string"
    )
      throw new Error("A run sessionId must be a string.");
    const context = sanitizeContext(request.context);
    const attachments = sanitizeAttachments(request.attachments);
    this.runs.start({
      requestId,
      prompt,
      sessionId:
        typeof request.sessionId === "string" ? request.sessionId : undefined,
      context,
      attachments,
    });
  }

  private cancel(requestId: string, reportUnknown: boolean): boolean {
    const cancelled = this.runs.cancel(requestId);
    if (!cancelled) {
      if (reportUnknown)
        this.sendError("unknown_request", "Unknown active request.", requestId);
      return false;
    }
    return true;
  }

  private send(message: RuntimeServiceMessage): void {
    if (this.closed) return;
    this.options.write(
      this.wireMode === "jsonrpc" ? jsonRpcMessage(message) : message,
    );
  }

  private sendError(
    code: RuntimeServiceErrorCode,
    message: string,
    requestId?: string,
  ): void {
    this.send({
      type: "error",
      ...(requestId ? { requestId } : {}),
      code,
      message,
    });
  }

  private unavailableCapability(
    requestId: string,
    capability: string,
  ): "continue" {
    this.sendError(
      "method_not_found",
      `${capability} is unavailable from this service.`,
      requestId,
    );
    return "continue";
  }

  private async hostResponse(
    requestId: string,
    load: () => Promise<RuntimeServiceResult>,
  ): Promise<void> {
    try {
      this.send({
        type: "response",
        requestId,
        result: await load(),
      });
    } catch (error) {
      this.sendError("internal_error", requestError(error), requestId);
    }
  }
}

export { runService } from "./protocol/stdio.js";

File: packages/neovim/lua/truss/protocol.lua
local M = {}
local version = require("truss.version")

M.protocol_version = version.protocol
M.client_version = version.plugin

local function default_encode(value)
  return vim.json.encode(value)
end

local function default_decode(value)
  return vim.json.decode(value)
end

local function schedule(callback)
  if vim.in_fast_event() then
    vim.schedule(callback)
  else
    callback()
  end
end

local Client = {}
Client.__index = Client

function Client.new(options)
  options = options or {}
  return setmetatable({
    command = options.command or "truss-cli",
    arguments = options.arguments or {},
    workspace = options.workspace or vim.fn.getcwd(),
    environment = options.environment,
    encode = options.encode or default_encode,
    decode = options.decode or default_decode,
    spawn = options.spawn or vim.system,
    on_event = options.on_event or function() end,
    on_lifecycle = options.on_lifecycle or function() end,
    on_approval = options.on_approval or function() end,
    on_diagnostic = options.on_diagnostic or function() end,
    on_exit = options.on_exit or function() end,
    pending = {},
    sequence = 0,
    stdout_buffer = "",
    initialized = false,
    capabilities = {},
    process = nil,
    active_request = nil,
    starting = false,
    start_callbacks = {},
  }, Client)
end

function Client:_next_id(prefix)
  self.sequence = self.sequence + 1
  return string.format("nvim-%s-%d", prefix, self.sequence)
end

function Client:_finish_pending(message)
  local request_id = message.id or message.requestId
  local callback = request_id and self.pending[request_id] or nil
  if not callback then
    return
  end
  self.pending[request_id] = nil
  local error_message = message.error
    and {
      message = message.error.message,
      code = message.error.data and message.error.data.code or message.error.code,
      data = message.error.data,
    }
    or (message.type == "error" and message or nil)
  callback(error_message, message)
end

function Client:_message(message)
  if type(message) ~= "table" then
    self.on_diagnostic("Truss service sent an invalid message.")
    return
  end
  if message.jsonrpc == "2.0" and message.method == "runtime/event" then
    local params = message.params or {}
    self.on_event(params.event, params.requestId)
  elseif message.jsonrpc == "2.0" and message.method == "run/lifecycle" then
    self.on_lifecycle(message.params or {})
  elseif message.jsonrpc == "2.0" and message.method == "approval/requested" then
    self.on_approval(message.params or {})
  elseif message.jsonrpc == "2.0" and (message.result or message.error) then
    self:_finish_pending(message)
  elseif message.type == "event" then
    self.on_event(message.event, message.requestId)
  elseif message.type == "lifecycle" then
    self.on_lifecycle(message)
  elseif message.type == "response" or message.type == "error" then
    self:_finish_pending(message)
  end
end

function Client:feed(chunk)
  if not chunk or chunk == "" then
    return
  end
  self.stdout_buffer = self.stdout_buffer .. chunk
  while true do
    local newline = self.stdout_buffer:find("\n", 1, true)
    if not newline then
      return
    end
    local line = self.stdout_buffer:sub(1, newline - 1)
    self.stdout_buffer = self.stdout_buffer:sub(newline + 1)
    if line ~= "" then
      local ok, message = pcall(self.decode, line)
      if ok then
        schedule(function()
          self:_message(message)
        end)
      else
        schedule(function()
          self.on_diagnostic("Could not decode Truss service output: " .. tostring(message))
        end)
      end
    end
  end
end

function Client:_write(message, callback)
  if not self.process then
    if callback then
      callback({ message = "Truss service is not running." })
    end
    return nil
  end
  local request_id = message.id or message.requestId
  if callback then
    self.pending[request_id] = callback
  end
  local encoded = self.encode(message) .. "\n"
  local ok, error_message = pcall(function()
    self.process:write(encoded)
  end)
  if not ok then
    self.pending[request_id] = nil
    if callback then
      callback({ message = tostring(error_message) })
    end
    return nil
  end
  return request_id
end

function Client:start(callback)
  if self.process and self.initialized then
    callback(nil, self.capabilities)
    return
  end
  table.insert(self.start_callbacks, callback)
  if self.starting then
    return
  end
  self.starting = true
  local function finish(error_message)
    self.starting = false
    if error_message and self.process then
      local process = self.process
      self.process = nil
      process:kill(15)
    end
    local callbacks = self.start_callbacks
    self.start_callbacks = {}
    for _, pending in ipairs(callbacks) do
      pending(error_message, error_message and nil or self.capabilities)
    end
  end
  local command = type(self.command) == "table" and vim.deepcopy(self.command)
    or { self.command }
  table.insert(command, "serve")
  vim.list_extend(command, self.arguments)
  local ok, process_or_error = pcall(self.spawn, command, {
    cwd = self.workspace,
    env = self.environment,
    stdin = true,
    text = true,
    stdout = function(error_message, data)
      if error_message then
        schedule(function()
          self.on_diagnostic(tostring(error_message))
        end)
      end
      self:feed(data)
    end,
    stderr = function(_, data)
      if data and data ~= "" then
        schedule(function()
          self.on_diagnostic(vim.trim(data))
        end)
      end
    end,
  }, function(result)
    schedule(function()
      self.process = nil
      self.initialized = false
      self.active_request = nil
      local message = string.format(
        "Truss service exited (%s).",
        tostring(result and result.code or "unknown")
      )
      for request_id, pending in pairs(self.pending) do
        self.pending[request_id] = nil
        pending({ message = message })
      end
      self.on_exit(result)
    end)
  end)
  if not ok then
    self.process = nil
    finish({ message = "Could not start Truss service: " .. tostring(process_or_error) })
    return
  end
  self.process = process_or_error
  local request_id = self:_next_id("initialize")
  self:_write({
    jsonrpc = "2.0",
    id = request_id,
    method = "initialize",
    params = {
      protocolVersions = { M.protocol_version },
      client = { name = "truss.nvim", version = M.client_version },
    },
  }, function(error_message, response)
    if error_message then
      finish(error_message)
      return
    end
    local result = response.result or {}
    if result.protocolVersion ~= M.protocol_version then
      finish({ message = "Truss service negotiated an unsupported protocol version." })
      return
    end
    self.initialized = true
    self.capabilities = result.capabilities or {}
    finish(nil)
  end)
end

function Client:request(method, params, callback)
  local handler = callback and function(error_message, response)
    callback(error_message, response and response.result or nil)
  end or nil
  return self:_write({
    jsonrpc = "2.0",
    id = self:_next_id("request"),
    method = method,
    params = params or {},
  }, handler)
end

function Client:run(input, callback)
  local request_id = self:_next_id("run")
  self.active_request = request_id
  self:_write({
    jsonrpc = "2.0",
    id = request_id,
    method = "run/start",
    params = {
      prompt = input.prompt,
      sessionId = input.session_id,
      context = input.context or {},
      attachments = input.attachments or {},
    },
  }, function(error_message, response)
    if self.active_request == request_id then
      self.active_request = nil
    end
    callback(error_message, response and response.result or nil)
  end)
  return request_id
end

function Client:cancel(callback)
  local target = self.active_request
  if not target then
    if callback then
      callback({ message = "No Truss run is active." })
    end
    return
  end
  self:_write({
    jsonrpc = "2.0",
    id = self:_next_id("cancel"),
    method = "run/cancel",
    params = { targetRequestId = target },
  }, callback)
end

function Client:approve(call_id, approved, callback)
  if not self.active_request then
    if callback then
      callback({ message = "No Truss run is active." })
    end
    return
  end
  self:request("approval/resolve", {
    callId = call_id,
    approved = approved,
  }, callback)
end

function Client:test_connection(callback)
  self:request("provider/test", {}, callback)
end

function Client:list_profiles(callback)
  self:request("profiles/list", {}, callback)
end

function Client:mcp_status(callback)
  self:request("mcp/status", {}, callback)
end

function Client:shutdown()
  if not self.process then
    return
  end
  self:_write({
    jsonrpc = "2.0",
    id = self:_next_id("shutdown"),
    method = "service/shutdown",
    params = {},
  })
  local process = self.process
  vim.defer_fn(function()
    if self.process == process then
      process:kill(15)
    end
  end, 250)
end

M.Client = Client

return M
